contracts/token/ERC721/ERC721.sol

```diff
@@ -221,40 +199,73 @@ abstract contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Er
      */
-    function _isApprovedOrOwner(address spender, uint256 tokenId) internal view virtual returns (bool) {
-        address owner = ownerOf(tokenId);
-        return (spender == owner || isApprovedForAll(owner, spender) || getApproved(tokenId) == spender);
+    function _getApproved(uint256 tokenId) internal view virtual returns (address) {
+        return _tokenApprovals[tokenId];
     }
 
     /**
-     * @dev Safely mints `tokenId` and transfers it to `to`.
+     * @dev Returns whether `spender` is allowed to manage `owner`'s tokens, or `tokenId` in
+     * particular (ignoring whether it is owned by `owner`).
      *
-     * Requirements:
-     *
-     * - `tokenId` must not exist.
-     * - If `to` refers to a smart contract, it must implement {IERC721Receiver-onERC721Received}, which is called upon a safe transfer.
+     * WARNING:
+     * - ownership is not checked.
+     * - spender = address(0) will lead to this function returning true in unexpected situations.
+     */
+    function _isApproved(address owner, address spender, uint256 tokenId) internal view virtual returns (bool) {
+        return owner == spender || isApprovedForAll(owner, spender) || getApproved(tokenId) == spender;
+    }
+
+    /**
+     * @dev Checks whether `spender` was approved by `owner` to operate on `tokenId` or is the owner of `tokenId`.
+     * Reverts if `spender` is not approved nor owner of `tokenId`.
      *
-     * Emits a {Transfer} event.
+     * WARNING:
+     * - ownership is not checked.
      */
-    function _safeMint(address to, uint256 tokenId) internal virtual {
-        _safeMint(to, tokenId, "");
+    function _checkApproved(address owner, address spender, uint256 tokenId) internal view virtual {
+        if (!_isApproved(owner, spender, tokenId)) {
+            revert ERC721InsufficientApproval(spender, tokenId);
+        }
     }
 
     /**
-     * @dev Same as {xref-ERC721-_safeMint-address-uint256-}[`_safeMint`], with an additional `data` parameter which is
-     * forwarded in {IERC721Receiver-onERC721Received} to contract recipients.
+     * @dev Transfers `tokenId` from its current owner to `to`, or alternatively mints (or burns) if the current owner
+     * (or `to`) is the zero address. Returns the owner of the `tokenId` before the update.
+     *
+     * The `operatorCheck` argument is optional. If the value passed is non 0, then this function will check that
+     * `operatorCheck` is either the owner of the token, or approved to operate on the token (by the owner).
+     *
+     * Emits a {Transfer} event.
      */
-    function _safeMint(address to, uint256 tokenId, bytes memory data) internal virtual {
-        _mint(to, tokenId);
-        if (!_checkOnERC721Received(address(0), to, tokenId, data)) {
-            revert ERC721InvalidReceiver(to);
+    function _update(address to, uint256 tokenId, address operatorCheck) internal virtual returns (address) {
```

Comment:
I am not convinced about either `operatorCheck` or `expectedOperator`. My issue with `expectedOperator` is that there isn't just one operator that we will check to be this one, there are potentially multiple operators and this is just the current one.

Perhaps the name should be something to indicate that the value is used for "authorization". `address auth`? It also gives a hint that it's an important variable that shouldn't be touched (in an override).

We could pair this with renaming `_is/checkApproved` to `_is/checkAuthorized` (instead of `_is/checkOperator` like I suggested elsewhere).

Reply:
I am not convinced about either `operatorCheck` or `expectedOperator`. My issue with `expectedOperator` is that there isn't just one operator that we will check to be this one, there are potentially multiple operators and this is just the current one.

Perhaps the name should be something to indicate that the value is used for "authorization". `address auth`? It also gives a hint that it's an important variable that shouldn't be touched (in an override).

We could pair this with renaming `_is/checkApproved` to `_is/checkAuthorized` (instead of `_is/checkOperator` like I suggested elsewhere).

Reply:
I like this idea really much! I would pick `checkAuthorized` from the options you gave. The rationale is that we've consistently named `_check` to the reverting functions (eg. `_checkOnERC721Received`) and `isAuthorized` might suggest a boolean returning value.

Reply:
So If I follow correctly:
- `_isApproved` → `_isAuthorised` (internal function)
- `_checkApproved` → `_checkAuthorised` (internal function)
- `operatorCheck` → `auth` (function parameter)

Do we all agree on that ?

Reply:
Yes but `authorized` is spelled with a "z" in American English which is what we use.

Reply:
done